src/Workspaces/Remote/Core/Telemetry/ApiUsageIncrementalAnalyzerProvider.cs

```diff
@@ -0,0 +1,211 @@
+
+            public void RemoveProject(ProjectId projectId)
+            {
+                lock (_reported)
+                {
+                    _reported.Remove(projectId);
+                }
+            }
+
+            public async Task AnalyzeProjectAsync(Project project, bool semanticsChanged, InvocationReasons reasons, CancellationToken cancellationToken)
+            {
+                lock (_reported)
+                {
+                    // to make sure that we don't report while solution load, we do this heuristic.
+                    // if the reason we are called is due to "document being added" to project, we wait for next analyze call.
+                    // also, we only report usage information per project once.
+                    // this telemetry will only let us know which API ever used, this doesn't care how often/many times an API
+                    // used. and this data is approximation not precise information. and we don't care much on how many times
+                    // APIs used in the same solution. we are rather more interested in number of solutions or users APIs are used.
+                    if (reasons.Contains(PredefinedInvocationReasons.DocumentAdded) ||
+                        _reported.Contains(project.Id))
+                    {
+                        return;
+                    }
+                }
+
+                var metadataSymbolUsed = new HashSet<ISymbol>();
+                foreach (var document in project.Documents)
+                {
+                    var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                    var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                    foreach (var operation in GetOperations(model, cancellationToken))
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
+
+                        if (metadataSymbolUsed.Count > Max)
+                        {
+                            // collect data up to max per project
+                            break;
+                        }
+
+                        // this only gather reference and method call symbols but not type being used.
+                        // if we want all types from metadata used, we need to add more cases 
+                        // which will make things more expansive.
+                        CollectApisUsed(operation, metadataSymbolUsed);
+                    }
+                }
+
+                var groupByAssembly = metadataSymbolUsed.GroupBy(s => s.ContainingAssembly);
+                var apiPerAssembly = groupByAssembly.Select(g => new
+                {
+                    // mark all string as PII (customer data)
+                    AssemblyName = new TelemetryPiiProperty(g.Key.Identity.Name),
+                    AssemblyVersion = g.Key.Identity.Version.ToString(),
```

Comment:
this should likely hash too right?  the version string contains things like the name of the assembly right? a nd version numbers are so specific they could likely identify assemblies pretty easily.

Reply:
pii is hashed. you can see what is hashed from PR description.

Reply:
Yes, but you didn't hash assembly version.  That's revealing information as it's not that hard to map a set of assembly versions back to a likely set of actual assemblies.  i.e. in your example:

`
"AssemblyVersion": "3.12.0.0",
`

that, along with other identifying versions would be very easy to use to get a highly likelihood clique of assemblies referenced.

Reply:
In other words, i tink all these properties should have TelemetryPiiProperty.  there shoudl only be exceptions where it has been investigated and signed off that there isn't a PII concern.

Reply:
The version is just a System.Version. You're thinking `Identity`, which is the assembly name + public key token + version + optional culture.

Reply:
>The version is just a System.Version

Yes.  But i remember from my PII training how just a few pieces of information could be used to easily identify things once you added some more info.  Think the "netflix deanonymization" example they used there :)  

Again, if there aren't pii concerns, that's fine.  I just want to make sure that this was explicitly thought through and hte information you get about a set of projects, their references ,and the versions you learn about that isn't felt to be a problem.

Reply:
The MS course had a key takeaway of: you'd be surprised how little data is needed to actually have a privacy violation.  "When in doubt, have the group expert here come check on things".

That's generally the rule i follow for anything i'm slightly nervous about where i know i'm not an expert :)

Reply:
> > The version is just a System.Version
> 
> Yes. But i remember from my PII training how just a few pieces of information could be used to easily identify things once you added some more info

That's a good point. I was more responding to your earlier comment:

> this should likely hash too right? the version string contains things like the name of the assembly right?

The only thing we shouldn't do is hash the compound string of assembly name and version, because this might make it hard for us to pregenerate the hashes.